akka-cluster/src/main/scala/akka/cluster/Cluster.scala

```diff
@@ -300,8 +300,10 @@ class Cluster(val system: ExtendedActorSystem) extends Extension {
    * The name of the [[akka.actor.ActorSystem]] must be the same for all members of a
    * cluster.
    */
-  def join(address: Address): Unit =
+  def join(address: Address): Unit = {
+    checkHostCharacters(address)
```

Comment:
should also be added to `akka.cluster.typed.Join` message (constructor check)

Reply:
@patriknw I wonder if this is needed (because it isn't so simple to call this new existing in the `Join` constructor from where it is in `Cluster` and in typed it is immediately checked in the call to untyped. If we let it get checked in untyped it is cleaner though I agree with you.

```scala
case Join(address) =>
  adaptedCluster.join(address) // executes checkHostCharacters
case JoinSeedNodes(addresses) =>
  adaptedCluster.joinSeedNodes(addresses) // executes checkHostCharacters
```

Reply:
since it's a message it will not be fail fast (at the sender side)

You can move `checkHostCharacters` to `Address`, or some other place that is accessible from akka-cluster-typed and akka-cluster. It doesn't have to live inside the `Cluster` class since it's just stateless validation logic.

Reply:
Sounds good, I originally thought it belonged in `Address` but didn't know if it was too leaky.

Reply:
yes, not good in Address since the error message talks about cluster joining. Put it in `Cluster` companion object.

Reply:
I removed the cluster part of the message and put it in Address for more reusability. If that makes sense.